lwc/src/test/scala/quasar/physical/s3/S3DataSourceSpec.scala

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.s3
+
+import slamdata.Predef._
+import S3DataSourceSpec._
+import cats.effect.IO
+import fs2.Stream
+import quasar.api.ResourceDiscoverySpec
+import quasar.api.{ResourceName, ResourcePath, ResourcePathType}
+import scalaz.{Foldable, Monoid, Id, ~>}, Id.Id
+import scalaz.syntax.applicative._
+import shims._
+import org.http4s.client.blaze.Http1Client
+import org.http4s.Uri
+
+final class S3DataSourceSpec extends ResourceDiscoverySpec[IO, Stream[IO, ?]] {
+  "the root of a bucket with a trailing slash is not a resource" >>* {
+    val root = ResourcePath.root() / ResourceName("")
+    discovery.isResource(root).map(_ must beFalse)
+  }
+
+  "the root of a bucket is not a resource" >>* {
+    val root = ResourcePath.root()
+    discovery.isResource(root).map(_ must beFalse)
+  }
+
+  "a prefix without contents is not a resource" >>* {
+    val path = ResourcePath.root() / ResourceName("prefix3") / ResourceName("subprefix5")
+    discovery.isResource(path).map(_ must beFalse)
+  }
+
+  "list nested children" >>* {
+    val path = ResourcePath.root() / ResourceName("dir1") / ResourceName("dir2") / ResourceName("dir3")
+
+    val listing = discovery.children(path)
+
+    listing.flatMap { list =>
+      list.map(_.compile.toList)
+        .getOrElse(IO.raiseError(new Exception("Could not list nested children under dir1/dir2/dir3")))
+        .map {
+          case List((resource, resourceType)) =>
+            resource must_= ResourceName("flattenable.data")
+            resourceType must_= ResourcePathType.resource
+        }
+    }
+  }
+
+  "read line-delimited and array JSON" >>* {
+    val ld = discoveryLD.evaluate(ResourcePath.root() / ResourceName("testData") / ResourceName("lines.json"))
+    val array = discovery.evaluate(ResourcePath.root() / ResourceName("testData") / ResourceName("array.json"))
```

Comment:
Do these two files exist in our S3 test bucket or are we mocking this?

Reply:
All of the data is in an actual bucket. It's not mocked. I think it's important we actually hit S3 in the test suite. It's currently in a bucket under my account, but I can change that to a company bucket soon.